### PR TITLE
UPC++/GASNet-EX 2022.3.0 update

### PR DESCRIPTION
**Pull request by bonachea**
This PR updates packages [UPC++](https://upcxx.lbl.gov) and [GASNet-EX](https://gasnet.lbl.gov)  from the [LBNL Pagoda group](https://go.lbl.gov/class) to the 2022.3.0 release.

This notably includes adding support for HPE Cray EX "Shasta" systems with Slingshot-10/11 networks, along with AMD ROCm HIP support and improvements to the NVIDIA CUDA support.

### CODE REVIEW

File: var/spack/repos/builtin/packages/upcxx/package.py [hunk 1]
```diff
@@ -22,7 +25,7 @@ def cross_detect():
     return 'none'
 
 
-class Upcxx(Package):
+class Upcxx(Package, CudaPackage, ROCmPackage):
     """UPC++ is a C++ library that supports Partitioned Global Address Space
     (PGAS) programming, and is designed to interoperate smoothly and
     efficiently with MPI, OpenMP, CUDA and AMTs. It leverages GASNet-EX to
```
Comment (PHHargrove) [inline var/spack/repos/builtin/packages/upcxx/package.py]: Please add ROCm/HIP to the list containing CUDA.

**Review by PHHargrove — COMMENTED** [review]
Configure args look right to me.  
However, I am wondering if we should be using a `libfabric` package to locate `fi_info` and to pass `--with-ofi-home=...`.   That depends strongly on whether or not Spack on Cray EX correctly provides such a package for the vendor-provided install.  Otherwise, my suggestion is clearly a bad one.

Also made one comment on a cosmetic issue in the upcxx package description.

**Comment by bonachea** [comment]
@PHHargrove wrote:
> I am wondering if we should be using a `libfabric` package to locate `fi_info` and to pass `--with-ofi-home=...`. That depends strongly on whether or not Spack on Cray EX correctly provides such a package for the vendor-provided install.

I considered this, however the current public Spack package for libfabric not surprisingly omits the proprietary CXI provider. I've also verified that `spack external find` is NOT smart enough to detect the Cray-provided libfabric install (in general it does a terrible job detecting Cray module packages). Taken together, this means adding `depends_on('libfabric')` to our spackage greatly increases the chances that someone will end up inadvertently building their own libfabric when installing our library, which _might_ be acceptable on SS-10, but is definitely the wrong outcome on SS-11. On balance I'd prefer to see the build fail and the user forced to type `module load libfabric`, especially since we only default-enable it on Cray.

**Comment by PHHargrove** [comment]
Not surprised to learn Spack doesn't do the right thing for HPE-provided libfabric.  I withdraw that suggestion.

**Comment by bonachea** [comment]
Upstreamed: https://github.com/spack/spack/pull/30393

**Pull request closed without merging**